contracts/providers/chainlink/ChainlinkPairAggregator.sol

```diff
@@ -38,15 +55,34 @@ contract ChainlinkPairAggregator is PairAggregatorInterface {
         return aggregator.latestRound();
     }
 
-    function getPreviousAnswer(uint256 roundsBack) external view returns (int256) {
+    function getPreviousTimestamp(uint256 roundsBack) external view returns (uint256) {
         uint256 latest = aggregator.latestRound();
-        require(roundsBack <= latest, "Not enough history");
-        return aggregator.getAnswer(latest - roundsBack);
+        require(roundsBack <= latest, "NOT_ENOUGH_HISTORY");
+        return aggregator.getTimestamp(latest - roundsBack);
     }
 
-    function getPreviousTimestamp(uint256 roundsBack) external view returns (uint256) {
+    /** Internal Functions */
+
+    function _getLatestAnswer() internal view returns (int256) {
+        return aggregator.latestAnswer();
+    }
+
+    function _getPreviousAnswer(uint256 roundsBack) internal view returns (int256) {
         uint256 latest = aggregator.latestRound();
-        require(roundsBack <= latest, "Not enough history");
-        return aggregator.getTimestamp(latest - roundsBack);
+        require(roundsBack <= latest, "NOT_ENOUGH_HISTORY");
+        return aggregator.getAnswer(latest - roundsBack);
+    }
+
+    function _normalizeDecimals(int256 value) internal view returns (int256) {
```

Comment:
Please can you explain the purpose of this function? Why is it that we need the answer to have at least tokenDecimals and maybe more?

Reply:
The values returned from Chainlink have different decimals (18,  8, others), and don't match with the token decimals. For example, the oracle LINK/USD has 8 decimals but the Link token has 18; but in case DAI/ETH, it has 18 decimals.
We use it to get the oracle price in "token decimals". So then, for example, we can get the inverse rate.
I have created unit tests where we mock the oracle price (based on real data from Ropste) and calculate the collateral info: LoansBaseGetCollateralInfoTest

For example, one of the unit tests is:
Price LINK/USD (Chainlink): 414000000 (8 decimals), but we need to use the inverse rate (USD/LINK).

Reply:
> We use it to get the oracle price in "token decimals".
If chainlink has 18 decimals and the token has 12, then the function gives the result 18 decimals - which is the same number of decimals as the chainlink oracle? How does the code then deal with that?

Reply:
The oracle price is only used in the LoansBase contracts (LoansBase._convertWeiToToken and LoansBase._convertTokenToWei). It deals in the same way that it does for the other scenarios.

There are unit tests for that scenario in the unit test test/base/LoansBaseGetCollateralInfoTest.js
- _4_usdc_eth_response_18_token_6_repay_0

This info is printed using the LoansBase.getCollateralInfo function (it uses LoansBase._convertTokenToWei internally)
In Lending Tokens: 50000000 USDC (6 decimals)
In Coll Tokens:    215645250000000000 WEI = 0.21564525 ETH
Oracle Price:      4312905000000000 => 1 USDC = 0.004312905 ETH

In that case, the "In Coll Tokens" is calculated 0.004312905 * 50 USDC = 0.21564525 ETH

The _convertXToY functions use a "whole lending token to calculate the output.

- _5_usdc_eth_response_18_token_6_repay_50
- _6_usdc_eth_response_18_token_6_repay_100

Reply:
Okay here is a scenario that I think it is wrong... Maybe I am misunderstanding though
- We have a market with USDC (6 decimals) as lending token
- SLINK (made up token... 8 decimals) as collateral token.
- The price oracle gives the price for 1 whole USDC in SLINK.
- The price oracle has 12 decimals

Let's say the price in for 1 USDC in whole SLINK is 3, so the price for 1 whole USDC in SLINK units is 300,000,000 (8 decimals). The Oracle uses 12 decimals and so has the price 3,000,000,000,000.

`LoansBase._convertTokenToWei` converts USDC units, to their unit SLINK cost

So lets say we call `LoansBase._convertTokenToWei(4000000)` i.e. 4 USDC units to SLINK units
```
uint256 aWholeLendingToken = 10**6
 // because tokenDecimals >= responseDecimals is not true, the next line does not alter the oracle's decimals
uint256 oneLendingTokenPriceWei = 3,000,000,000,000
uint256 tokenValue = 4000000.mul(3,000,000,000,000).div(10**6);
                                 = 12,000,000,000,000
```
So this function says that 4 USDC units is 12 trillion SLINK units, which is 120,000 whole SLINK. However we know that it should be 12 SLINK for 4 USDC.

The reason for this difference in output is because the oracle has 4 more decimals than the token.

Reply:
Great catch! I have added 30 unit tests (total) to cover your scenario and all the possible ones.